bindings/python/src/lib.rs

```diff
@@ -30,18 +30,41 @@ use pyo3::types::PyBytes;
 use pyo3::types::PyDict;
 
 mod asyncio;
+mod layers;
 
 use crate::asyncio::*;
 
 create_exception!(opendal, Error, PyException);
 
-fn build_operator(scheme: od::Scheme, map: HashMap<String, String>) -> PyResult<od::Operator> {
+fn add_layers(mut op: od::Operator, layers: Vec<layers::Layer>) -> PyResult<od::Operator> {
```

Comment:
To avoid additional costs, apply the layers before calling `finish`.

Reply:
You can't really do that because `OperatorBuilder.layer` changes the return type, so
`builder = builder.layer(..)` doesn't compile.

```rust
error[E0308]: mismatched types=======> ] 240/241: opendal-python
  --> bindings/python/src/lib.rs:45:74
   |
40 |     mut builder: od::OperatorBuilder<impl od::raw::Accessor>,
   |                  -------------------------------------------
   |                  |                   |
   |                  |                   this type parameter
   |                  expected due to this parameter type
...
45 |             layers::Layer::Retry(layers::RetryLayer(inner)) => builder = builder.layer(inner),
   |                                                                          ^^^^^^^^^^^^^^^^^^^^ expected type parameter `impl od::raw::Accessor`, found struct `RetryAccessor`
   |
   = note: expected struct `OperatorBuilder<impl od::raw::Accessor>`
              found struct `OperatorBuilder<opendal::layers::retry::RetryAccessor<impl od::raw::Accessor>>`


error[E0308]: mismatched types=======> ] 240/241: opendal-python
  --> bindings/python/src/lib.rs:47:27
   |
40 |     mut builder: od::OperatorBuilder<impl od::raw::Accessor>,
   |                  -------------------------------------------
   |                  |                   |
   |                  |                   this type parameter
   |                  expected due to this parameter type
...
47 |                 builder = builder.layer(inner);
   |                           ^^^^^^^^^^^^^^^^^^^^ expected type parameter `impl od::raw::Accessor`, found struct `ImmutableIndexAccessor`
   |
   = note: expected struct `OperatorBuilder<impl od::raw::Accessor>`
              found struct `OperatorBuilder<opendal::layers::immutable_index::ImmutableIndexAccessor<impl od::raw::Accessor>>`


error[E0308]: mismatched types=======> ] 240/241: opendal-python
  --> bindings/python/src/lib.rs:50:27
   |
40 |     mut builder: od::OperatorBuilder<impl od::raw::Accessor>,
   |                  -------------------------------------------
   |                  |                   |
   |                  |                   this type parameter
   |                  expected due to this parameter type
...
50 |                 builder = builder.layer(inner);
   |                           ^^^^^^^^^^^^^^^^^^^^ expected type parameter `impl od::raw::Accessor`, found struct `ConcurrentLimitAccessor`
   |
   = note: expected struct `OperatorBuilder<impl od::raw::Accessor>`
              found struct `OperatorBuilder<opendal::layers::concurrent_limit::ConcurrentLimitAccessor<impl od::raw::Accessor>>`
```

Reply:
> You can't really do that because `OperatorBuilder.layer` changes the return type, so
> `builder = builder.layer(..)` doesn't compile.

The result type should be `Operator`

Reply:
The error is before calling `finish()`, the problem is you can't assign the return type of `OperatorBuilder.layer` to a variable in a for-loop.